### ISSUE TITLE
On the calculation of Xt partial loss in Zt

### ISSUE THREAD

**Issue by CCCCCCCCCdut** [Shark-NLP/DiffuSeq#26]
https://github.com/Shark-NLP/DiffuSeq/blob/main/diffuseq/gaussian_diffusion.py#:~:text=terms%5B%22mse%22%5D%20%3D%20mean_flat((target%20%2D%20model_output)%20**%202)
https://github.com/Shark-NLP/DiffuSeq/blob/main/diffuseq/utils/nn.py#:~:text=def%20mean_flat(,.shape))))
Dear author, since you keep the Xt part of Zt unchanged (it is anchored as X0 after each calculation), when calculating the MSE loss of Lt-1, the Xt part is 0, will the loss result be wrong? I tried the mean_flat function and it counted 0 into the number. thank you!

**Comment by summmeer** [comment]
Hi,
It is true that the $x_t$ part is 0, referring to Eq.(2) in the paper. There is still $y_t$ part for the mse loss when training.

**Comment by CCCCCCCCCdut** [comment]
> Hi, It is true that the xt part is 0, referring to Eq.(2) in the paper. There is still yt part for the mse loss when training.

1. Yes, what you mainly train is the loss in yt, but the X part entered into the mean_flat function is 0, and the function will also count the number of X, resulting in a smaller loss. For example, Zt = [0,0,0,2,8], where 0 is the part of MSE_X, it actually calculates the average of (2+8)/5 (including the number of X's and Y's) instead of calculating the average of (2+8)/2 (the number of Y's).
2. Is the structure of your Transformer the same as that in DiffuSeq? Thanks!

**Comment by summmeer** [comment]
Hi,
1. You're right, it is more rigorous. For the whole sequence, the total length is fixed to 128, and the length of x would not vary much, so I think this loss is bearable.
2. The embedding learns the word embedding of each token, just the same as other pretrain language models. The word embedding here also corresponds to the semantic meaning of the token, e.g. similar words have closer vector cosine scores.
3. There is no doubt that larger data helps better language modeling capacity, but I am not sure how DiffuSeq works on a small set of data.
4. The transformer architecture (layers, hidden dims, etc) and the training methods (vlb loss, optimizer, learning rate annealing) are similar to Diffusion-LM.

**Comment by CCCCCCCCCdut** [comment]
> Hi,
> 
> 1. You're right, it is more rigorous. For the whole sequence, the total length is fixed to 128, and the length of x would not vary much, so I think this loss is bearable.
> 2. The embedding learns the word embedding of each token, just the same as other pretrain language models. The word embedding here also corresponds to the semantic meaning of the token, e.g. similar words have closer vector cosine scores.
> 3. There is no doubt that larger data helps better language modeling capacity, but I am not sure how DiffuSeq works on a small set of data.
> 4. The transformer architecture (layers, hidden dims, etc) and the training methods (vlb loss, optimizer, learning rate annealing) are similar to Diffusion-LM.

Thank you for your answer. For the second Embedding section. There is a Embedding(EMB) for mapping W to X0 by consulting the source code, but there is also a word_embedding in the Bert model itself, which shares the weight with lm_head. What's the difference? By referring to your code you will map X to Xt using word_embedding in Bert to implement the sample when sampling. So what role do the W to X0 EMBs trained previously in _token_discrete_loss play in the sampling process?

**Comment by summmeer** [comment]
Hi,
I guess your "Embedding(W)" refers to randomly initialized:
https://github.com/Shark-NLP/DiffuSeq/blob/bdc8f0adbff22e88c8530d1f20c3c7589c061d40/basic_utils.py#L73

In fact, we didn't use it in the end, both for training and sampling. This code is the legacy of the previous work and we leave it as a possible interface that you can initialize word embedding using your own weights (e.g. Glove embedding). We only use `model.word_embedding`.

**Comment by CCCCCCCCCdut** [comment]
> 

Thank you for your patience,
Good luck!

**Comment by CCCCCCCCCdut** [comment]
> Hi, I guess your "Embedding(W)" refers to randomly initialized:
> 
> https://github.com/Shark-NLP/DiffuSeq/blob/bdc8f0adbff22e88c8530d1f20c3c7589c061d40/basic_utils.py#L73
> 
> In fact, we didn't use it in the end, both for training and sampling. This code is the legacy of the previous work and we leave it as a possible interface that you can initialize word embedding using your own weights (e.g. Glove embedding). We only use `model.word_embedding`.

Thank you for your patience,
Good luck!
